Better organise the documentation, document the policy engine, the authorisation process and the scopes (#2637)
matrix-org/matrix-authentication-service: docs/topics/authorization.md

# Authorization and sessions

The main job of the authentication service is to grant access to resources to clients, and to lets resources know who is accessing them.
In less abstract terms, this means that the service is responsible for issuing access token and let the homeserver (and other services) introspect those access tokens.

## How access tokens work

In MAS, the access token is an opaque string for which the service has metadata associated with it.
An access token has:

- a subject, which is the user the token is issued for
- a list of [scopes](../reference/scopes.md)
- a client for which the token is issued
- a timeframe for which the token is valid

> On a single token, those metadata are immutable: they don't change over time.

Reviewer: ```suggestion
On a single token, metadata is immutable: they don't change over time.
```

## How Synapse behaves

When an incoming request is made to Synapse, it will introspect the access token through the Matrix Authentication Service.
This is using a standard OAuth 2.0 introspection request ([RFC 7662]).

Out of this request, Synapse will care about the following:

- the `active` field, which tells if the token is valid or not
- the `sub` field, which tells which user the token is issued for. This is an opaque string, and Synapse saves the mapping between the Matrix user ID and the subject of the token in its own database
- in case Synapse doesn't know the presented subject, it will look at the `username` field, which it will use as the localpart for the user as fallback
- the `scope` field, which tells which scopes are granted to the token. More specifically, it will look for the following scopes:
  - [`urn:matrix:org.matrix.msc2967.client:api:*`], which grants broad access to the whole Matrix C-S API
  - [`urn:matrix:org.matrix.msc2967.client:device:AABBCC`], which encodes the Matrix device ID used by the client
  - [`urn:synapse:admin:*`], which grants access to the Synapse admin API

> It's important to understand that when Synapse delegates the authentication to MAS, a lot of user attributes are not managed by Synapse anymore.

Reviewer: ```suggestion
It's important to understand that when Synapse delegates authentication to MAS, Synapse no longer manages many user attributes.
```

This exists as a compatibility layer for clients that don't yet support OAuth 2.0, but has some restrictions compared to the way those sessions behaved in Synapse.

When a client presents a compatibility access token to Synapse, MAS will make it look like to Synapse as if the token had the following scopes:

- [`urn:matrix:org.matrix.msc2967.client:api:*`]
- [`urn:matrix:org.matrix.msc2967.client:device:AABBCC`]

Which corresponds to the broad access to the Matrix C-S API and the device ID of the client, as one would expect from the legacy login API.
One important missing scope is [`urn:synapse:admin:*`], which means that the client won't have access to the Synapse admin API.

This is the case even if the user has the `can_request_admin` attribute set to `true`, and this is by design:
the legacy login API doesn't have a way to request specific scopes, and we don't want to grant admin access to all clients that have a compatibility session.
This was the case in the past with Synapse, as the admin status was set on the user itself, but this is not the case anymore with MAS.

## OAuth 2.0 sessions

Modern clients are expected to use OAuth 2.0 to authenticate with the homeserver.

> In OAuth 2.0/OIDC, there are multiple way to start an OAuth 2.0 session called grants.

Reviewer: ```suggestion
In OAuth 2.0/OIDC, there are multiple ways to start an OAuth 2.0 session called grants.
```
